### ISSUE TITLE
Review Unihan database changes for Unicode 15.1

### ISSUE THREAD

**Issue by kipcole9** [elixir-unicode/unicode_unihan#8]
[Unicode 15.1](https://unicode.org/versions/Unicode15.1.0/) is due for release in September 2023.  There are some changes that may require an update to this code in this library.

> Documentation was added for CJK Unified Ideographs Extension I and for 6 new provisional properties. 7 old provisional properties were removed. The syntax, list of sources, and/or descriptions were updated for the kIRG_GSource, kIRG_KSource, and kIRG_KPSource properties. Syntax and descriptions were also updated for several other properties, including kRSUnicode.

**Comment by kipcole9** [comment]
I have updated the Unihan database to [Unicode 15.1](https://unicode.org/versions/Unicode15.1.0)  in the `v15.1` git branch.  Unicode 15.1 contains 627 new CJK characters.  While the download works, the decode process now raises an exception:

```elixir
kip@Kips-MBP unicode_unihan % mix unicode.unihan.download
Compiling 1 file (.ex)

13:18:05.642 [info] Downloaded "https://www.unicode.org/Public/UCD/latest/ucd/Unihan.zip" to "/Users/kip/Development/unicode_unihan/_build/dev/lib/unicode_unihan/priv/unihan.zip"

13:18:06.504 [info] Downloaded "https://www.unicode.org/Public/UCD/latest/ucd/CJKRadicals.txt" to "/Users/kip/Development/unicode_unihan/_build/dev/lib/unicode_unihan/priv/cjk_radicals.txt"
** (ArgumentError) errors were found at the given arguments:

  * 1st argument: not a textual representation of an integer

    (erts 14.0.2) :erlang.binary_to_integer("C255'", 16)
    (unicode_unihan 0.3.0) lib/unicode_unihan/utils.ex:246: Unicode.Unihan.Utils.decode_metadata/3
    (unicode_unihan 0.3.0) lib/unicode_unihan/utils.ex:88: anonymous fn/5 in Unicode.Unihan.Utils.parse_file/2
    (elixir 1.15.4) lib/map.ex:957: Map.get_and_update/3
    (unicode_unihan 0.3.0) lib/unicode_unihan/utils.ex:86: anonymous fn/3 in Unicode.Unihan.Utils.parse_file/2
    (elixir 1.15.4) lib/enum.ex:4387: anonymous fn/3 in Enum.reduce/3
    (elixir 1.15.4) lib/stream.ex:1654: Stream.do_element_resource/6
    (elixir 1.15.4) lib/enum.ex:4387: Enum.reduce/3
```

This suggests there is a different encoding in some part of the data. Will you be able to take a look at this?

**Comment by kipcole9** [comment]
Ping?

**Comment by jkwchui** [comment]
Hi Kip.  The earlier msg came at a particularly busy time.  Sorry.

There are changes that need to be kept up-to-date, but adding/removing fields are relatively low priority.  What breaks the build is something odd in how radicals are structured: there is a new annotation for radicals, that makes use of `''` as the decoration.  182 was previously

```
182; 2FB5; 98A8 #⾵, 風 (traditional radical, traditional ideograph)
182'; 2EDB; 98CE #⻛, 风 (simplified radical, simplified ideograph)
```
to which the following entry is added:
```
182''; ; 322C4 # [some related glyph](https://zi-hi.com/sp/uni/322C4) in Ext Plane H
```
The same happens for 208, 210, 211, 212, and 213.

I'm looking at the notes in [#38-35](https://www.unicode.org/reports/tr38/tr38-35.html), but I don't see where it indicated the change to `CJKradicals.txt`.  I can't find what the `''` indicates; the docs (in Annex #44) seems to say they just "are" ("Added several entries for radical variants using the two apostrophes convention.").  We had previously 

1. bifurcated the rows into "traditional" (no prime) and "simplified" (prime), and 
2. assumed there is a radical unicode *and* an ideograph unicode

These two assumptions are both challenged.  I am not sure what is the right way to handle this, but I feel that patching to make the parse fitting the shape is the wrong approach.  Let me try to dig a little more and see if I can find meaning behind the `''`.

**Comment by kipcole9** [comment]
@jkwchui all good, thanks for taking a look into this.  What you're seeing is quite common (in my experience) with Unicode. Something change and they're well documented, some things change and they are conformant to the spec but the spec is unclear. And some things just change!

I'm pretty sure that whatever approach you choose to take will be acceptable to the people who find the library useful. And of course no special rush intent ended.

Let me know how I can help?

**Comment by jkwchui** [comment]
The new variant is (probably) Japanese equivalent of the radical (e.g., "亀").  I have made a commit to the 15.1 branch that parses these into `:Hanj`, so `Unicode.Unihan.Radical.radical(182, :all)` gives

```
%{
        Hans: %{
          radical_character: 11995,
          radical_number: 182,
          unified_ideograph: 39118
        },
        Hant: %{
          radical_character: 12213,
          radical_number: 182,
          unified_ideograph: 39080
        },
        Hanj: %{
          radical_character: nil,
          radical_number: 182,
          unified_ideograph: 205508
        }
      }
```

This thus extends but not break the existing API.  Two issues:

1. (minor) rendering of the radical goes through `Unicode.Unihan.to_string/1`, and it raises when `:nil` is provided.  This may or may not be the right behaviour here.
2. (**major**) `mix test` is broken with

```
Parsing the Unihan database (this may take a few seconds).
????????????????????????????????

  0) Unicode.Test: failure on setup_all callback, all tests have been invalidated
     ** (ArgumentError) errors were found at the given arguments:

       * 1st argument: not a textual representation of an integer
 ```

Can you take a look at this?

**Comment by kipcole9** [comment]
@jkwchui, Its taken me way too long to get back to this but its a cold rainy day here to perfect for the task. I've merged your changes from last year and moved a little bit forward in the `unicode_15.1` branch. 

However this is one bug I can't resolve because it appears there has been an update to the `unihan_fields` data with a new field `kVietnameseNumeric` that isn't in that file. And I don't know how to source of generate it.

```elixir
kip@Kips-iMac-Pro unicode_unihan % mix unicode.unihan.download

14:21:11.658 [info] Downloaded "https://www.unicode.org/Public/UCD/latest/ucd/Unihan.zip" to "/Users/kip/Development/unicode_unihan/_build/dev/lib/unicode_unihan/priv/unihan.zip"

14:21:13.581 [info] Downloaded "https://www.unicode.org/Public/UCD/latest/ucd/CJKRadicals.txt" to "/Users/kip/Development/unicode_unihan/_build/dev/lib/unicode_unihan/priv/cjk_radicals.txt"
** (RuntimeError) Unknown field :kVietnameseNumeric found for "9"

    (unicode_unihan 0.3.0) lib/unicode_unihan/utils.ex:272: Unicode.Unihan.Utils.maybe_split_value/3
    (unicode_unihan 0.3.0) lib/unicode_unihan/utils.ex:254: Unicode.Unihan.Utils.decode_metadata/3
    (unicode_unihan 0.3.0) lib/unicode_unihan/utils.ex:88: anonymous fn/5 in Unicode.Unihan.Utils.parse_file/2
    (elixir 1.17.1) lib/map.ex:957: Map.get_and_update/3
    (unicode_unihan 0.3.0) lib/unicode_unihan/utils.ex:86: anonymous fn/3 in Unicode.Unihan.Utils.parse_file/2
    (elixir 1.17.1) lib/enum.ex:4423: anonymous fn/3 in Enum.reduce/3
    (elixir 1.17.1) lib/stream.ex:1717: Stream.do_element_resource/6
    (elixir 1.17.1) lib/enum.ex:4423: Enum.reduce/3
```

If you let me know where to find this data, or how to generate it, I'll keep going on the fixes.

I'll admit there is a little bit of selfish self-interest here too. I'm hoping you can guide me on how to sort ideographs - Im going to work on unicode collations again very soon. :-)

**Comment by kipcole9** [comment]
I've made some more progress. I wrote a new module `Unicode.Unihan.Property` that has functions to download and parse the HTML definitions of Unihan properties and saves them as an `.etf` file.  I updated Unihan to the Unicode 15.1 version and then updated the tests to match.

I *think* its ok to go now subject to your validation of course.

Comments, suggestions, updates welcome of course.

**Comment by kipcole9** [comment]
One concern I have is that the properties can change with each release. Therefore the property data and the Uniihan release files have to be in sync. And therefore I suspect the download process needs to download both the Unihan database and the Unihan properties.

I think on this for tomorrow but I expect I'll merge those workflows together into a single process.

**Comment by kipcole9** [comment]
> I expect I'll merge those workflows together into a single process

I've implemented that integration and believe this is now ready for release which, subject to any feedback, I'll publish this week.

**Comment by kipcole9** [comment]
I've published [Unicode Unihan version 0.4.0](https://hex.pm/packages/unicode_unihan/0.4.0) with the following changelog entry:

### Enhancements

* Updated to [Unicode 15.1](https://www.unicode.org/versions/Unicode15.1.0/)

* Property information is autogenerated from [TR38 UNICODE HAN DATABASE (UNIHAN](https://www.unicode.org/reports/tr38/#AlphabeticalListing).

* Diagnostic output uses `Logger.info/2` rather than `IO.puts/1`
